hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1207,32 +1217,24 @@ def minimize_floats(self, chooser):
         anything particularly meaningful for non-float values.
         """
 
-        ex = chooser.choose(
-            self.examples,
-            lambda ex: (
-                ex.label == DRAW_FLOAT_LABEL
-                and len(ex.children) == 2
-                and ex.children[1].length == 8
+        leaf = chooser.choose(self.leaves, lambda leaf: leaf.ir_type == "float")
```

Comment:
Right now, we call `Float.shrink` on any float node (+ some short circuiting behavior in `Float`). The previous behavior skipped `Float.shrink` on floats that happened to also be integers:

```python
        buf = self.shrink_target.buffer
        b = buf[u:v]
        f = lex_to_float(int_from_bytes(b))
        b2 = int_to_bytes(float_to_lex(f), 8)
        if b == b2 or self.consider_new_buffer(buf[:u] + b2 + buf[v:]):
```

I think the `if b == b2` condition there is equivalent to checking `if is_simple(f)`. i.e., don't try to float-shrink float nodes if they are already integers. The comments in `lexical.py` in [this commit](https://github.com/HypothesisWorks/hypothesis/commit/47f2adc47953619c9a93cfb3dd7578937db3429e#diff-5044b2f8ebb16009a425e4c08a026c87b6611d9986f3b0c16f92b5329c057508L81-L89) give some further insight: 

```python
        # This floating point number can be represented in our simple format.
        # So we try converting it to that (which will give the same float, but
        # a different encoding of it). If that doesn't work then the float
        # value of this doesn't unambiguously give the desired predicate, so
        # this approach isn't useful. If it *does* work, then we're now in a
        # situation where we don't need it, so either way we return here.
        if is_simple(f):
            self.incorporate_float(f)
            return
```

To be honest, I'm not sold that we don't want to shrink float nodes as floats even if they happen to currently be integers. We mostly defer to the `Integer` shrinker anyway, and it also seems to make logical sense to keep all float-shrink-logic in the same place instead of jumping around. But I'm probably missing a good historical reason for it to be done this way. It's easy to put this condition back either way.

Reply:
I think the intuition here is that we consider integer-valued floats simpler than all non-integer-valued floats, and so once we get to an integer-value we might as well use the machinery we have for shrinking integers.

You could probably convince me that a different ordering is better, but that's how I understand the status quo.

Reply:
ok, I think that makes sense to me. I'll add this safeguard back. I was thinking about the case where shrinking an integer-valued float is faster if you allow it to try intermediate float values. But `Integer` probably gets you close enough to the minimum and then you can finish it off with a final `Float` pass. 

I'd like to take a detailed look at the shrink ordering at some point. I saw a few weird things like values in (0, 1) being ordered after almost everything else; `assert float_to_lex(sys.float_info.max) < float_to_lex(0.5)`. But for another day! 

I think Hypothesis' shrinker is so well tuned that I shouldn't make changes without profiling or being very confident in them first.

Reply:
Values between zero and one: not integers, and we order kinda by number of decimal places, so.  And the largest finite float is an integer, which means "simpler" than 0.5!  (I did say I'm not that attached to the ordering)